### ISSUE TITLE
Maps with identically named Sources overwrite the older Source's data

### ISSUE THREAD

**Issue by neoeno** [mapbox/mapbox-gl-js#3411]
**mapbox-gl-js 0.26.0**:

### Steps to Trigger Behavior

[Reproduction demo which probably makes it clearer than my description](http://jsbin.com/yoduruqeyi/edit?html,output). Tried in Chrome, FF. Note it takes a few seconds to set up, so wait for the red notice to disappear. [Here's a video of what I see.](http://recordit.co/YvMzW8YQIj)

In abstract:

1. Load `map1`
2. Load a new square feature Source named `fred` into `map1`
3. Load `map2` (ensuring `map1` has fully loaded)
4. Load a new triangle feature Source also named `fred` into `map2`
5. Zoom out in `map1` (guessing this triggers a re-simplification & some render work)

__Expected Behavior:__ `map1` shows a square

__Actual Behavior:__ `map1` shows a triangle

### Technical investigation

I've looked into this a little and it seems like it's around the fact that [the worker pool is global](https://github.com/mapbox/mapbox-gl-js/blob/master/js/global_worker_pool.js#L4) (i.e. each map uses the same pool), and each worker stores a copy of the data it is sent in [an object indexed by the source name](https://github.com/mapbox/mapbox-gl-js/blob/15960ff04674a46a98bccb421f81e5886a24924c/js/source/geojson_worker_source.js#L83) — hence, if two maps use the same source name, one will overwrite the other and the next time the worker is asked to render it uses the latest one it has received.

As such, maybe this is intended behaviour — there is a [sentence in the docs that seems to allude to this](https://github.com/mapbox/mapbox-gl-js/blob/7e8cf5d15ae932d54c45219ebd51cfbaebd213b2/js/ui/map.js#L658):

```javascript
    /**
     * Adds a source to the map's style.
     *
     * @param {string} id The ID of the source to add. Must not conflict with existing sources.
```

But I guess I wouldn't expect that to mean 'existing sources, including in other maps' — since the map instances are all separate and as a user I'm not thinking about there being globals behind the scenes.

### Actions

Happy to do the work on any of these or some other solution — just give me a steer on what your ideal would be.

Possibilities:

1. Keep it as-is
2. Change the docs to clarify the matter
3. Index the worker cache object by something including a unique ID to distinguish maps

Thanks for all your good work — we all love mapbox-gl over here! ✨

**Comment by jfirebaugh** [comment]
This is a fantastic bug report. Thank you for investigating!

The worker creates a separate `GeoJSONWorkerSource` for each map. The reason the cache is nevertheless shared between maps is that `GeoJSONWorkerSource` is using a single object value from the prototype object rather than initializing `this._geoJSONIndexes = {}` in the constructor.

**Comment by neoeno** [comment]
Ah — I missed that! Just tried a fix out too and it fixes my original issue. Thanks!

If you'd like a PR and a test I'm happy to take a look at that tomorrow? Or if you'd prefer to make the change in-house and save on PR-admin I perfectly understand.

**Comment by jfirebaugh** [comment]
A PR would be great, thank you!

**Comment by mourner** [comment]
I accidentally found this and fixed as a part of #3408. It's great that ES6 classes don't support prototype properties — makes it easier to find bugs like this. :)
